for task up to date check both status and sources

harelwa opened the pull request:
> Thanks for your pull request, we really appreciate contributions!
>
> Please understand that it may take some time to be reviewed.
>
> Also, make sure to follow the [Contribution Guide](https://taskfile.dev/contributing/).

Closes #840 

- **Disclaimer:** I'm not a `go` programmer. so if there is a more "idiomatic go" way of doing what i did, do tell.
- **Test:** As `TestStatusChecksum` checks exactly what I think should be checked in this case, I've added a `status` to the tested `Taskfile`. Without my changes to `status.go` this test will fail when running on the modified `Taskfile`

ghostsquad commented on status.go:
```diff
@@ -33,13 +33,16 @@ func (e *Executor) isTaskUpToDate(ctx context.Context, t *taskfile.Task) (bool,
 		return false, nil
 	}
```
while we are here, there's really no point in this if statement. It's redundant, and only clutters this function.

ghostsquad commented on status.go:
```diff
@@ -53,11 +56,13 @@ func (e *Executor) isTaskUpToDate(ctx context.Context, t *taskfile.Task) (bool,
 		}
 		if !isUpToDate {
-			return false, nil
+			isUpToDateChecker = false
```
does `checker.IsUpToDate()` have a side effect of writing the checksum file?

harelwa replied:
indeed

harelwa replied:
https://github.com/go-task/task/blob/d4ed7c3cfc2408bf0978b9d693ff226e3b9c0707/internal/status/checksum.go#L49

ghostsquad reviewed (changes requested):
I think this still needs a test

harelwa commented:
> ~I think this still needs a test~ maybe a _new_ test can be written to catch this case? We want to ensure both cases are valid, so please don't modify existing tests.

it can be the same test running on two difference `Taskfile`s - the prev one + the one here

harelwa commented:
> ~I think this still needs a test~ maybe a _new_ test can be written to catch this case? We want to ensure both cases are valid, so please don't modify existing tests.

\+ if you have noticed something the existing test does not cover for this case, please share

harelwa commented:
> > ~I think this still needs a test~ maybe a _new_ test can be written to catch this case? We want to ensure both cases are valid, so please don't modify existing tests.
> 
> it can be the same test running on two difference `Taskfile`s - the prev one + the one here

i do think such an addition is somewhat redundant though, don't you ?

ghostsquad commented:
The test is verifying the behaviour of a single piece of functionality in isolation. This change affects the combination of features. Thus, technically we want a permutation of tests.
- status - OK
- status - failed
- source - OK
- source - failed
- status - OK & source OK
- ... etc.

harelwa commented:
> The test is verifying the behaviour of a single piece of functionality in isolation. This change affects the combination of features. Thus, technically we want a permutation of tests.
> 
> * status - OK
> * status - failed
> * source - OK
> * source - failed
> * status - OK & source OK
> * ... etc.

I'd not re-test `status, as `TestStatus` already covers some of these [subtests](https://go.dev/blog/subtests).

I'd rather re-use/ride `TestStatusChecksum` as it tests the desired functionality correctness.

Following your remark, I'd also rather not "combine" tests like I did.

[thus] I have added subtests to `TestStatusChecksum` and added a separate subtest accordingly, leaving the previous one unchanged.

harelwa commented:
thank you @ghostsquad for the kind and fast review. and approval.

Are we good to merge ?

pd93 commented:
@andreynering @ghostsquad @abrahammurciano This hasn't been properly tested. It seems to have broken checksums for any tasks that don't declare `status` or `sources`.

On `d4ed7c3`

```sh
❯ task --dir ./testdata/env
task: [local] echo "GOOS='$GOOS' GOARCH='$GOARCH' CGO_ENABLED='$CGO_ENABLED'" > local.txt
task: [global] echo "FOO='$FOO' BAR='$BAR' BAZ='$BAZ'" > global.txt
```

On `master` (`6e346de`)

```sh
❯ task --dir ./testdata/env
task: Task "default" is up to date
```

See fix here: #1037
